File: packages/scene-wgpu/src/phongWgpuMeshMaterialRenderer.ts
import type { LinearColor } from '@flighthq/materials';
import { unpackColorToLinear } from '@flighthq/materials';
import { getWgpuRenderStateRuntime } from '@flighthq/render-wgpu';
import type {
  Camera,
  Material,
  MeshGeometry,
  PhongMaterial,
  SceneLightBlock,
  SceneRenderProxy,
  WgpuMeshMaterialRenderer,
  WgpuRenderState,
} from '@flighthq/types';
import { PhongMaterialKind } from '@flighthq/types';

import type { WgpuClassicDefineKey } from './wgpuClassicPrelude';
import { bindWgpuClassicSurface, ensureWgpuClassicPipeline } from './wgpuClassicPrelude';
import { registerWgpuMeshMaterialRenderer } from './wgpuMeshMaterialRegistry';
import {
  beginWgpuMeshDraw,
  drawWgpuMeshSubset,
  hasWgpuMaterialTexture,
  writeWgpuFrameUniform,
} from './wgpuMeshPipeline';

// The built-in classic Phong forward-lit mesh-material renderer (WgpuMeshMaterialRenderer for
// PhongMaterialKind) — the WGSL mirror of phongGlMeshMaterialRenderer. Lambert diffuse plus a
// reflection-vector specular lobe: bind selects the classic uber-shader's `phong` variant for the
// material's alpha mode + color format, writes the shared Frame uniform (camera position AND view-
// projection, since the specular term is view-dependent, plus the packed light block), binds the
// pipeline + Frame group (beginWgpuMeshDraw), then binds the material's linear diffuse + specular
// colors and shininess at group(2). draw issues the indexed draw. See registerPhongWgpuMaterial to
// install it.
export const phongWgpuMeshMaterialRenderer: WgpuMeshMaterialRenderer = {
  bind(
    state: WgpuRenderState,
    material: Readonly<Material> | null,
    lights: Readonly<SceneLightBlock>,
    camera: Readonly<Camera>,
  ): void {
    const stateRuntime = getWgpuRenderStateRuntime(state);
    const pass = stateRuntime.renderPass;
    if (pass === null) return;

    const phong = material as Readonly<PhongMaterial> | null;
    const format = stateRuntime.currentColorFormat ?? state.format;
    const pipeline = ensureWgpuClassicPipeline(state, defineKeyForMaterial(phong), format);
    writeWgpuFrameUniform(state, camera, lights);

    let group: GPUBindGroup;
    if (phong === null) {
      group = bindWgpuClassicSurface(state, pipeline, FALLBACK_MATERIAL, WHITE, WHITE, 32, 0.5, null, null, null);
    } else {
      unpackColorToLinear(_diffuse, phong.diffuse);
      unpackColorToLinear(_specular, phong.specular);
      group = bindWgpuClassicSurface(
        state,
        pipeline,
        phong,
        _diffuse,
        _specular,
        phong.shininess,
        phong.alphaCutoff,
        phong.diffuseMap,
        phong.specularMap,
        phong.normalMap,
      );
    }

    beginWgpuMeshDraw(state, pipeline);
    pass.setBindGroup(2, group);
  },

  draw(state: WgpuRenderState, proxy: Readonly<SceneRenderProxy>, geometry: Readonly<MeshGeometry>): void {
    drawWgpuMeshSubset(state, proxy, geometry);
  },
};

// Registers the built-in Phong renderer for PhongMaterialKind on this state. Opt-in (no top-level
// side effect); call once per WgpuRenderState before drawWgpuScene so meshes with PhongMaterials draw.
export function registerPhongWgpuMaterial(state: WgpuRenderState): void {
  registerWgpuMeshMaterialRenderer(state, PhongMaterialKind, phongWgpuMeshMaterialRenderer);
}

// The feature define key for a Phong material: the fixed `phong` lighting model plus the alpha-mask +
// double-sided flags and which optional maps are present (mirrors phongGlMeshMaterialRenderer).
function defineKeyForMaterial(material: Readonly<PhongMaterial> | null): WgpuClassicDefineKey {
  return {
    alphaMaskEnabled: material !== null && material.alphaMode === 'mask',
    doubleSided: material !== null && material.doubleSided,
    hasDiffuseMap: material !== null && hasWgpuMaterialTexture(material.diffuseMap),
    hasNormalMap: material !== null && hasWgpuMaterialTexture(material.normalMap),
    hasSpecularMap: material !== null && hasWgpuMaterialTexture(material.specularMap),
    lightingModel: 'phong',
  };
}

const _diffuse: LinearColor = [0, 0, 0, 0];
const _specular: LinearColor = [0, 0, 0, 0];
const WHITE: LinearColor = [1, 1, 1, 1];
const FALLBACK_MATERIAL = {} as Readonly<PhongMaterial>;

File: crates/flighthq-scene-wgpu/src/phong_wgpu_mesh_material_renderer.rs
//! The built-in classic Phong forward-lit mesh-material renderer — the WGSL mirror
//! of `phong_gl_mesh_material_renderer`.
//!
//! Ports `@flighthq/scene-wgpu` `phongWgpuMeshMaterialRenderer.ts`. Lambert diffuse
//! plus a reflection-vector specular lobe: `bind` selects the classic uber-shader's
//! `phong` variant, writes the shared Frame uniform (camera position + the packed
//! light block), and binds the material's linear diffuse + specular colors and
//! shininess at group(2); `draw` issues the indexed draw. Maps are deferred on wgpu.

use flighthq_materials::phong_material_kind;
use flighthq_render_wgpu::WgpuRenderState;
use flighthq_types::camera::Camera;
use flighthq_types::classic_material::PhongMaterial;
use flighthq_types::material::Material;
use flighthq_types::pbr_material::MaterialAlphaMode;
use flighthq_types::scene_render::{SceneLightBlock, SceneRenderProxy};

use crate::wgpu_classic_prelude::{
    WgpuClassicDefineKey, WgpuClassicLightingModel, bind_wgpu_classic_surface,
    ensure_wgpu_classic_pipeline,
};
use crate::wgpu_mesh_material_registry::{
    WgpuMeshMaterialRenderer, register_wgpu_mesh_material_renderer,
};
use crate::wgpu_mesh_pipeline::{
    draw_wgpu_mesh_material_subset, unpack_color_to_linear_f32, write_wgpu_frame_uniform,
};
use crate::wgpu_scene_runtime::{WgpuMeshUpload, WgpuSceneRuntime};

/// The built-in classic Phong forward-lit mesh-material renderer.
pub struct PhongWgpuMeshMaterialRenderer;

impl WgpuMeshMaterialRenderer for PhongWgpuMeshMaterialRenderer {
    fn bind(
        &self,
        state: &mut WgpuRenderState,
        scene: &mut WgpuSceneRuntime,
        material: Option<&dyn Material>,
        lights: &SceneLightBlock,
        camera: &Camera,
    ) {
        let phong = material.and_then(downcast_phong);
        let format = state.runtime.current_color_format.unwrap_or(state.format);
        let cache_key = ensure_wgpu_classic_pipeline(state, scene, &define_key(phong), format);
        scene.active_pipeline_key = Some(cache_key.clone());
        write_wgpu_frame_uniform(state, scene, camera, lights);

        let material_key = material.map_or_else(phong_material_kind, |m| m.kind());
        let (diffuse, specular, shininess, alpha_cutoff) = match phong {
            Some(m) => (
                unpack_color_to_linear_f32(m.diffuse),
                unpack_color_to_linear_f32(m.specular),
                m.shininess,
                m.alpha_cutoff,
            ),
            None => (WHITE, WHITE, 32.0, 0.5),
        };
        bind_wgpu_classic_surface(
            state,
            scene,
            &cache_key,
            material_key,
            &diffuse,
            &specular,
            shininess,
            alpha_cutoff,
        );
        scene.active_material_key = Some(material_key);
    }

    fn draw(
        &self,
        state: &mut WgpuRenderState,
        scene: &mut WgpuSceneRuntime,
        proxy: &SceneRenderProxy,
        upload: &WgpuMeshUpload,
    ) {
        draw_wgpu_mesh_material_subset(state, scene, proxy, upload);
    }
}

/// Returns the singleton Phong renderer instance.
pub fn phong_wgpu_mesh_material_renderer() -> PhongWgpuMeshMaterialRenderer {
    PhongWgpuMeshMaterialRenderer
}

/// Registers the built-in Phong renderer for the Phong material kind on this scene
/// runtime. Opt-in (no top-level side effect).
pub fn register_phong_wgpu_material(state: &mut WgpuRenderState, scene: &mut WgpuSceneRuntime) {
    register_wgpu_mesh_material_renderer(
        state,
        scene,
        phong_material_kind(),
        Box::new(PhongWgpuMeshMaterialRenderer),
    );
}

fn define_key(material: Option<&PhongMaterial>) -> WgpuClassicDefineKey {
    WgpuClassicDefineKey {
        alpha_mask_enabled: material.is_some_and(|m| m.alpha_mode == MaterialAlphaMode::Mask),
        double_sided: material.is_some_and(|m| m.double_sided),
        has_diffuse_map: false,
        has_normal_map: false,
        has_specular_map: false,
        lighting_model: WgpuClassicLightingModel::Phong,
    }
}

fn downcast_phong(material: &dyn Material) -> Option<&PhongMaterial> {
    (material as &dyn core::any::Any).downcast_ref::<PhongMaterial>()
}

const WHITE: [f32; 4] = [1.0, 1.0, 1.0, 1.0];

#[cfg(test)]
mod tests {
    use super::*;
    use crate::wgpu_scene_runtime::create_wgpu_scene_runtime;

    mod define_key {
        use super::*;

        #[test]
        fn selects_the_phong_lighting_model() {
            assert_eq!(
                define_key(None).lighting_model,
                WgpuClassicLightingModel::Phong
            );
        }
    }

    mod register_phong_wgpu_material {
        use super::*;

        #[test]
        fn registers_the_renderer_under_the_phong_kind() {
            let mut scene = create_wgpu_scene_runtime();
            scene.material_registry.insert(
                phong_material_kind(),
                Box::new(PhongWgpuMeshMaterialRenderer),
            );
            assert!(scene.material_registry.contains_key(&phong_material_kind()));
        }
    }
}
